drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,417 @@
+     * to the configuration that was actually used. A caller relying on a
+     * specific Schmitt trigger configuration must hence check if the
+     * configuration was applied as requested.
+     */
+    bool schmitt_trigger;
+    /**
+     * @brief   Initial value of the output
+     *
+     * Ignored if @ref gpio_conf_t::state is set to @ref GPIO_INPUT or
+     * @ref GPIO_DISCONNECT. If the pin was previously in a high impedance
+     * state, it is guaranteed to directly transition to the given initial
+     * value.
+     */
+    bool initial_value;
+    /**
+     * @brief   Strength of the pull up/down resistor
+     *
+     * This value is ignored when @ref gpio_conf_t::pull is configured to
+     * @ref GPIO_FLOATING. A call to @ref gpio_ll_init will select the pull
+     * resistor closet to the specified value and update this field to the
+     * actually used value.
+     */
+    gpio_pull_strength_t pull_strength;
+    /**
+     * @brief   Drive strength of the GPIO
+     *
+     * This value is ignored when @ref gpio_conf_t::state is configured to
+     * @ref GPIO_INPUT or @ref GPIO_DISCONNECT. A call to @ref gpio_ll_init
+     * will select the drive strength closet to the specified value and update
+     * this field to the actually used value.
+     */
+    gpio_drive_strength_t drive_strength;
+} gpio_conf_t;
+#endif
+
+/**
+ * @brief   Initialize the given GPIO pin as specified
+ *
+ * @param[in]       port    port the pin to initialize belongs to
+ * @param[in]       pin     number of the pin to initialize
+ * @param[in,out]   conf    configuration to apply
+ *
+ * @retval  0           success
+ * @retval  -ENOTSUP    GPIO state or pull register configuration not supported
+ *
+ * @note    If the configuration of the Schmitt trigger, the drive strength, or
+ *          the pull resistor strength are not supported, the closest supported
+ *          value will be chosen instead and `0` is returned. The value of these
+ *          fields in @p conf will be updated to the actually chosen value.
+ */
+int gpio_ll_init(gpio_port_t port, uint8_t pin, gpio_conf_t *conf);
+
+/**
+ * @brief   Get the current value of all GPIO pins of the given port as bitmask
+ *
+ * @param[in] port      port to read
+ *
+ * @return              state of all pins as bitmask
+ *
+ * @note    The value of unconfigured pins and pins configured as @ref
+ *          GPIO_DISCONNECT or @ref GPIO_OUTPUT_PUSH_PULL implementation
+ *          defined. For pins configured as @ref GPIO_OUTPUT_OPEN_DRAIN must
+ *          reflect what is read from the pin, so that a software I2C bus with
+ *          support for clock stretching can be implemented.
+ */
+static inline uword_t gpio_ll_read(gpio_port_t port);
+
+/**
+ * @brief   Perform an `reg |= mask` operation on the I/O register of the port
+ *
+ * @note    Only pins configured as @ref GPIO_OUTPUT_PUSH_PULL or as
+ *          @ref GPIO_OUTPUT_OPEN_DRAIN will be affected.
```

Comment:
It turned out that ATmega MCUs use the output register to configure the pull-ups for inputs. So it would cause overhead to filter out pins not configured as outputs. I'll change this to a warning that behavior is implementation defined